src/engine/condition.ts

```diff
@@ -1,23 +1,36 @@
 import * as reducers from './reducers';
 import * as matchers from './matchers';
 import { PageView, EngineCondition } from '../../types';
+import { dotProduct } from '../utils';
 
-const createCondition = (condition: EngineCondition) => (
-  pageViews: PageView[]
+const createCondition = <T>(condition: EngineCondition) => (
+  pageViews: PageView<T>[]
 ): boolean => {
   const { filter, rules } = condition;
   const filteredPageViews = filter.queries
     .map((query) => {
       return pageViews.filter((pageView) => {
-        const queryFeatures = pageView.features[query.property];
-        return queryFeatures.some((v) => query.value.indexOf(v) !== -1);
+        if (query.property === 'keywords') {
+          const queryFeatures = pageView.features[query.property];
+          return (queryFeatures || []).some(
+            (v) => query.value.indexOf(v) !== -1
+          );
+        } else if (query.property === 'topicModel') {
+          const queryFeatures = pageView.features[query.property];
+          return queryFeatures
+            ? dotProduct(queryFeatures.vector, query.value.vector) >
+                query.value.threshold
+            : false;
+        } else {
+          return true;
+        }
       });
     })
     .flat();
 
   const ruleResults = rules.map((rule) => {
     // TODO: allow other reducers...
```

Comment:
Yeah, you can refine the type using conditionals -- the previous commit that added the topic model check in the rules instead of the query had something like that set up for this. We could probably also use generics if it needs to be more polymorphic